update client-sessions for Windows

seanmonstar opened the pull request:
this has a new version of node-proxy, that uses
node-gyp and can therefore compile on Windows

This addresses 1 part of #1769.

shane-tomlinson commented:
@seanmonstar - I cannot correctly get this to install on my local machine or on remote ephemeral instances.

locally, I am using node v0.6.15, npm version v1.1.16, which is what we are using on the ephemeral instances.

local:

```
[14:22][browserid-remotes(client-sessions-windows)] $ npm install
npm http GET https://registry.npmjs.org/client-sessions/0.0.8
npm http GET https://github.com/downloads/lloyd/node-statsd/0509f85.tgz
npm http 304 https://registry.npmjs.org/client-sessions/0.0.8
npm http 200 https://github.com/downloads/lloyd/node-statsd/0509f85.tgz
npm http GET https://registry.npmjs.org/mersenne
npm http GET https://github.com/benadida/cookies/tarball/d2f0f0b3
npm http GET https://registry.npmjs.org/node-proxy/0.6.0
npm http 304 https://registry.npmjs.org/mersenne
npm http 304 https://registry.npmjs.org/node-proxy/0.6.0
npm http 200 https://github.com/benadida/cookies/tarball/d2f0f0b3
npm ERR! error rolling back client-sessions@0.0.8 Error: ENOTEMPTY, rmdir '/Users/stomlinson/development/browserid-remotes/node_modules/client-sessions'

npm ERR! Unsupported
npm ERR! Not compatible with your version of node/npm: node-proxy@0.6.0
npm ERR! Required: {"node":">=0.6","npm":">= 1.1.17"}
npm ERR! Actual:   {"npm":"1.1.16","node":"0.6.15"}
npm ERR! 
npm ERR! System Darwin 11.4.0
npm ERR! command "node" "/Users/stomlinson/development/nvm/v0.6.15/bin/npm" "install"
npm ERR! cwd /Users/stomlinson/development/browserid-remotes
npm ERR! node -v v0.6.15
npm ERR! npm -v 1.1.16
npm ERR! code ENOTSUP
npm ERR! message Unsupported
npm ERR! errno {}
npm ERR! 
npm ERR! Additional logging details can be found in:
npm ERR!     /Users/stomlinson/development/browserid-remotes/npm-debug.log
npm not ok
[14:23][browserid-remotes(client-sessions-windows)] $ node --version
v0.6.15
[14:26][browserid-remotes(client-sessions-windows)] $ npm --version
1.1.16
[14:26][browserid-remotes(client-sessions-windows)] $ 

```

remote:

```
[app@ip-10-144-18-16 ~]$ node --version
v0.6.15
[app@ip-10-144-18-16 ~]$ npm --version
1.1.16
[app@ip-10-144-18-16 ~]$ 
```

seanmonstar commented:
Well, crap, you're right. I hadn't noticed since I had updated my npm version. npm 1.1.7 was when `node-gyp` was merged in, which is how cross-platform native modules are compiled going forward.

Do we say that npm itself should be updated to .7? npm 1.1.7 wasn't bundled with node until node 0.6.13. We require at least 0.6.7.

Other option, is that when we upgrade to 0.8, this will upgrade npm to a compatible version also.

jrgm commented:
We should probably update the ami used for ephemeral deploys to match the version of node that we run in production. Developers should update their own node/npm to match. Here are the values on stage:

```
[jrgm@host ~]$ node --version
v0.6.17
[jrgm@host ~]$ node `which npm` --version
1.1.21
```

jrgm commented:
We should probably also update any documentation that says anything less than node 0.6.17

seanmonstar commented:
depends on #2250

lloyd commented:
#2250 is fixed.  rebase against current dev and let travis vet this PR for us.

shane-tomlinson commented:
using node v0.6.17 and npm 1.1.21, once I rebase from dev, the backend tests stall out on add-email-with-assertion-test.js

shane-tomlinson commented:
Follow on, further debug info:

```
[10:15][browserid-remotes(client-sessions-windows)] $ vows -v --spec tests/add-emaiwith-assertion-test.js 
The "sys" module is now called "util". It should have a similar interface.
vows bin argv [ 'tests/add-email-with-assertion-test.js' ]
vows bin options {
    reporter: 'dot-matrix',
    matcher: /.*/
}
vows runner loading [ '/Users/stomlinson/development/browserid-remotes/tests/add-email-with-assertion-test' ]
Setting NODE_ENV to test_json to test with the local JSON database
To test with a local mysql database, setup mysql and set NODE_ENV to test_mysql
vows runner running auth-with-assertion 
♢ auth-with-assertion

  specifying an ephemeral database
    ✓ should work

  opening the database
    ✓ should work fine

  run the server server should be running
    ✓ server is running

  generating a keypair
    ✓ succeeds

  generting a certificate
    ✓ works swimmingly

  generating an assertion
    ✓ succeeds

  adding this email via assertion
    ✗ fails if not authenticated
      » expected 400,
    got  403 (===) // add-email-with-assertion-test.js:114

  stage an account
    ✗ works
      » expected 200,
    got  403 (===) // add-email-with-assertion-test.js:128
```

seanmonstar commented:
I see, so this came because we skipped 2 versions, from 0.0.6 to 0.0.8. In 0.0.7, this landed: https://github.com/benadida/node-client-sessions/commit/f21761b43ff4f456b531ec5fe190765ffce7f0de

By going into `loadFromCookie` and setting `force_reset = true;` explicitly, then the browserid tests pass. All of client-sessions tests pass, so we need to add a test that mimics what we're doing in browserid.

I don't yet fully know how this code works, @fmarier would you be able to add more context?

fmarier commented:
I'm not sure what context would be useful, but the reasoning behind the change is in this bug:

  https://github.com/benadida/node-client-sessions/issues/14

By setting force_reset to True in loadFromCookie, I guess you're effectively reverting that patch.

seanmonstar commented:
OK, I found it. It was because undefined values were returning null, and browserid checks for undefined.

Fixed in benadida/node-client-sessions#17

seanmonstar commented:
While that change has been merged upstream, it hasn't been published as a new version in npm. Should I wait for that, or change the version in browserid's package.json to the commit sha?

shane-tomlinson commented:
@seanmonstar - @benadida has published node-client-sessions v0.0.9, which includes your fixes. Can this PR be updated to use 0.0.9 instead?

seanmonstar committed:
update client-sessions to 0.0.9

this has a new version of node-proxy, that uses
node-gyp and can therefore compile on Windows

seanmonstar commented:
@shane-tomlinson rocking the brand new 0.0.9, and includes an updated lockdown (since lockdown didn't exist when this PR was first opened).

shane-tomlinson commented:
ship it!

r+, works locally, works on ephemeral instances.

I will merge on Monday.
